### ISSUE TITLE
Text in search box should not be saved

### ISSUE THREAD

**Issue by braykov** [Jermolene/TiddlyWiki5#1971]
If I search for text, modify the tiddler I found and then saved the wiki - the search box contents is saved too.
I think it should not be.
1) It doesn't make any sense to keep it saved (at least to me it doesn't)
2) typing in the search box does not indicate a wiki modification, i.e. the save icon does not turn red, so why clicking it saves it, then?
Latest stable 5.1.9

**Comment by tobibeer** [comment]
Knowing that the search input is stored in `$:/temp/search`, this should already get you where you want: [States And Temporary Tiddlers @ tb5](http://tobibeer.github.io/tb5/#States%20And%20Temporary%20Tiddlers), so perhaps consider closing this ticket.

However, if you are asking to change the default, in particular for search, leave it open.

**Comment by braykov** [comment]
It is ok that there is a current workaround, but I prefer to spur a discussion.
I do believe in my points above. TW is a work in progress. Everyone would agree it should act as people intuitively expect it too. So let's hear them.

**Comment by tobibeer** [comment]
I'd very much prefer the general rule to be...

> **To persist any state or temporary tiddlers one has to be explicit about it.**
>    &raquo; possibly via namespace prefixes

Rather than the current situation being...

> *To* **not** *persist any state or temporary tiddlers one has to be explicit about it.*

**Comment by braykov** [comment]
Exactly my point.

**Comment by pmario** [comment]
imo the `$:/config/SaverFilter` tiddler already contains a parameter `-[prefix[$:/temp]]`, that should prvent all temporary tiddlers from saving. .. so seems to be a bugn

**Comment by tobibeer** [comment]
Oh, ok. Then I'll withdraw [the above](https://github.com/Jermolene/TiddlyWiki5/issues/1971#issuecomment-143952709) seeing as how it was indeed already changed during my abstinence these past few months.

@braykov, is it possible you are actually having an overwritten `$:/config/SaverFilter` that didn't receive a core update at some point?

**Comment by braykov** [comment]
This is my SaveFilter, and I haven't ever toched it
`[all[]] -$:/HistoryList -$:/StoryList -$:/Import -$:/isEncrypted -$:/UploadName -[prefix[$:/state]] -[prefix[$:/temp]]`
If this turns out to be a bug, then even better.

**Comment by tobibeer** [comment]
@braykov, can you upload an MTC somewhere, perhaps attached to a google groups post?

**Comment by braykov** [comment]
I don't see the point. There is nothing special about it. Anyone can just download the empty.html from the site and try it for themselves. It by default remembers the searched text.

**Comment by tobibeer** [comment]
Ah, ok. Just did that, and, after all the back and forth, I confirm that downloading an **empty.html** produces the reported behavior both on tiddlywiki.com as well as on tiddlywiki.com/prerelease

**Comment by tobibeer** [comment]
@Jermolene, I guess the problem lies in the core using the filter defined in **saveTiddlerFilter** in the "template" [$:/core/save/all](https://github.com/Jermolene/TiddlyWiki5/blob/master/core/templates/save-all.tid) rather than [$:/config/SaverFilter](https://github.com/Jermolene/TiddlyWiki5/blob/master/core/wiki/config/SaverFilter.tid) when actually saving.

This makes me wonder whether **$:/config/SaverFilter** is currently only used for setting the wiki dirty or not. But then how come it is separate from the above templates?

Further looking into **$:/core/save/all** I can see that one can additionally declare a variable called **publishFilter** to hold yet another filter expression based on which to further extend or shrink the output set, which makes me wonder even more what the purpose of this is as compared to **$:/config/SaverFilter**?

I get the feeling that one of the two needs to either go, or perhaps not be a variable to be set via global macro, because why? — I'm assuming that happened because there was no other convenient way to concatenate a filter in **$:/config/SaverFilter**, that concatenation of parameters isn't quite as straight forward as one might hope it would be.

Eventually, I'm assuming that all core [saving templates](https://github.com/Jermolene/TiddlyWiki5/tree/master/core/templates) should respect [$:/config/SaverFilter](https://github.com/Jermolene/TiddlyWiki5/blob/master/core/wiki/config/SaverFilter.tid) and perhaps do away with any seemingly superfluous **publishFilter**.

To me it feels like the saving mechanism needs some thorough consolidation.

**Comment by Jermolene** [comment]
State tiddlers such as those starting `$:/state` or `$:/temp` are included when saving the wiki. That includes the current search term in `$:/temp/search`.

The general idea is to try to preserve the wiki state across restarts, so that it comes up in the state in which it was left. If it's not the behaviour you want then at the moment the only workaround is to edit the filter in `$:/core/save/all`.

> This makes me wonder whether $:/config/SaverFilter is currently only used for setting the wiki dirty or not. But then how come it is separate from the above templates?

''$:/config/SaverFilter'' is indeed only used for detecting the dirty state of the wiki.

> Eventually, I'm assuming that all core saving templates should respect $:/config/SaverFilter and perhaps do away with any seemingly superfluous publishFilter.

Isn't the issue here just that we have different configuration tiddlers for the dirty state vs. the saving state? It's actually intentional. The idea was to be able to have tiddlers that were excluded from the dirty calculations but were still included in any subsequent. It's the cause of the behaviour observed in the OP:  typing into the search box doesn't make the wiki dirty, but the search term is still saved.

''publishFilter'' was added quite recently as a way for plugins to make changes to the tiddlers that are saved by default. It's not a very robust/useful mechanism since only one plugin at a time can change it.

**Comment by felixhayashi** [comment]
@Jermolene 

Now I am completely confused :( You once told me that `$:/temp` prefixed tiddlers are never stored.

>> can't we do it like /tmp in linux: delete $:/temp/ during bootup! that would be sweet! Then no cleanup is necessary.

>Hi Felix,
>By default, $:/temp tiddlers are not saved when the wiki is saved. […]

>https://groups.google.com/d/msg/tiddlywikidev/yuQB1KwlKx8/SCZqEhYPiN4J

Now you say:

> State tiddlers such as those starting $:/state or $:/temp are included when saving the wiki. That includes the current search term in $:/temp/search.

> The general idea is to try to preserve the wiki state across restarts, so that it comes up in the state in which it was left. If it's not the behaviour you want then at the moment the only workaround is to edit the filter in $:/core/save/all.

But this is highly problematic. I use `$:/temp` in a real temporary sense: to store temporary data that only makes sense for the current session. At the start of the next session, I expect that whole "polution" of the previous session to be gone. I also do no cleanup for stuff in the `$:/temp` path. This means there is tons of temp data lying around that bloats the wiki and nobody is aware of?

While nodejs people do not have to worry as `$:/temp` behaves as expected (because it is not synced), this is bad news for standalone users.

I any case. I really need a place to store temporary data that only exists for a session (like linux' `/tmp/`. This is a critical aspect of a system in my opinion and what is the purpose of the `$:/temp/` convention anyway, if it is not treated as such – or only acknowledged in nodejs-mode.

-Felix

**Comment by felixhayashi** [comment]
For example I just opened up one of my single page mode wikis and there is tons of stuff that looks like this:

```
$:/temp/felixhayashi/tiddlymap/holders/3565d0fd-143c-48f9-ab8b-04b0f14ada34
$:/temp/felixhayashi/tiddlymap/holders/40640c9f-9d8d-499e-ada6-c4e0a9ed9e8c
$:/temp/felixhayashi/tiddlymap/holders/440bc590-9aab-4c0f-8783-deeef8460803
$:/temp/felixhayashi/tiddlymap/holders/49f5080c-b97b-4b0a-97f6-ad51690bb20a
$:/temp/felixhayashi/tiddlymap/holders/4c3847d7-0329-4b40-b730-aca7ff186f8f
$:/temp/felixhayashi/tiddlymap/holders/51d527fb-fea3-4c7d-83b8-aba9da5632f4
$:/temp/felixhayashi/tiddlymap/holders/5291e201-c215-456a-a4b4-0588b12276f1
```

**Comment by Jermolene** [comment]
Hi @felixhayashi 

> Now I am completely confused :( You once told me that $:/temp prefixed tiddlers are never stored over sessions.

It was my confusion originally. I think the behaviour changed a few times over the early days of TW5 development, and I'd forgotten where it ended up.

I made the same mistake as @pmario in the thread that you referenced, forgetting which template is used for which purpose. [As an aside, I do wonder about adding a brief `description` field to the core templates; it would surely save a lot of people a lot of digging and headscratching].

Anyhow, the question now is whether we should save `$:/state` and/or `$:/temp` tiddlers. I tend to think that we should save `$:/state`, but not `$:/temp`.

**Comment by felixhayashi** [comment]
> It was my confusion originally. I think the behaviour changed a few times over the early days of TW5 development, and I'd forgotten where it ended up.

:D 

> I tend to think that we should save $:/state, but not $:/temp.

:+1: That would be great! I mean this is how `$:/temp` works with nodejs because of the non-syncing and it would be great if this were also the case for standalone!

**Comment by Evolena** [comment]
> Anyhow, the question now is whether we should save $:/state and/or $:/temp tiddlers. I tend to think that we should save $:/state, but not $:/temp.

I agree that it would be much more understandable.

However, we still have an inconsistency: $:/state/popup are NOT saved.

**Comment by pmario** [comment]
>Anyhow, the question now is whether we should save $:/state and/or $:/temp tiddlers. I tend to think that we should save $:/state, but not $:/temp.

+1

>$:/state/popup are NOT saved.

IMO should not be saved, because this caused a lot of trouble. imo with alpha.

**Comment by Evolena** [comment]
I don't say they shouldn't, I point to this exception so that it is not forgotten in the discussion ;)

**Comment by tobibeer** [comment]
Allow me to repeat this:

**the general rule to be...**

> **To persist** any state or temporary tiddlers one **needs to be explicit** about it.
>    &raquo; possibly via namespace prefixes

**rather than the current situation being...**

> *To* **not** *persist any state or temporary tiddlers one needs to be explicit about it.*

This resolves most, if not all of the confusion.

But I would still very much prefer that "dirty" and "needs saving" (as in "is going to be saved") mean very much the same. I can't see a situation where I would want both. Either we're dirty or we're not. It's not just that we only don't want the save button to be red so much. We don't want the wiki to be dirty when we search something, open a popup, toggle a slider... and if we want to actually persist that state, there will be a way to be specific about that, if only by not using tiddlers under a temporary namespace.

Other than that, neither `$:/temp/` nor `$:/state/` should be persisted ...both of which focus on a single-user, single-visitor paradigm, plus the assumption that the basic user expectation is to find the wiki exactly as it was, meaning in every little state'ish detail ...which is wrong, imho.

You don't expect your desktop or web application to open with the "edit menu" open or with the "settings tab" showing as of your last session, unless that was a specific URL / place you actually visited.. but not in all its final state, only in its initial presentation.

**Comment by pmario** [comment]
My opinion and a proposal:

>Other than that, neither $:/temp/ nor $:/state/ should be persisted, imho, both of which focus on a single-user, single-visitor paradigm, plus the assumption that the basic user expectation is to find the wiki exactly as it was, meaning in every little state'ish detail.

IMO that's not true for everyone

If state variables should not be saved, they need to be prefixed with `$:/temp`. Everything else should be saved. If there are cases where we don't want `$:/state` to be saved, they should be renamed to `$:/temp` 

There are different opinions which elements should use `$:/state` and `$:/temp` and there will be no possibility to have the right default settings for everyone. So what ever is the default it will be wrong.

eg: With my workflow, I want persistent `$:/state` variables, while I'm working with the wiki. But I want default settings if I publish them in "read only" mode. 

##Proposal

 - `$:/temp` variables will not be saved
 - `$:/state` variables will be saved 
 - `$:/config/states`  is a json tiddler that is used to initialize state tiddlers at startup.
   - if the `$:/config/states` tiddler does not exist, all states will be saved. (current behaviour)
   - if the `$:/config/states` tiddler does exit, but is empty all state tiddlers will be deleted.
   - if the `$:/config/states` contains json data, all state tiddlers will be deleted and the json data will be used to initialize the configured states. 

Advantage of this mechanism: 

- No problem with backwards compatibility, since we can design the new behaviour accordingly
- We have a consistent behaviour, that is easy to describe.
   - `$:/temp` variables will be temporary, everything else is persistent.
- It's super easy to initialize state tiddlers
   - clear all states with an empty config tiddler
   - create a config tiddler if you need special behaviour for your public wikis
   - rename / or remove the config tiddler, if you need to persist your states 

This configuration tiddler also has several advantages:

- it's easy to create, with advanced search: filters eg: `[prefix[$:/state/]]` and export it.
- We can drag this config tiddler from one TW to an other and create the exact same state. 
   - this is very powerful if we need to help users. 
   - eg: open the tiddler info panel and select the "fields" tab, is difficult to describe. 
      - import this config tiddler is simple. 

what do you think?

**Comment by braykov** [comment]
@pmario you seem to have thought of everything. I tend to agree with this approach.
This feature of TW looks very important to me. So if implemented this way I could only add the following:
* This should be well documented somewhere (or in the future help)
* A place in the Control Panel, where you could modify the `$:/config/states` tiddler and also have a check-box to have enabled (state tiddlers will be deleted) or disabled (all states will be saved). I just think such kind of behavioral mechanism really deserves easy access via GUI configuration tool because it is important to how saving works.

**Comment by tobibeer** [comment]
@pmario 

> If state variables should not be saved, they need to be prefixed with $:/temp. Everything else should be saved. If there are cases where we don't want `$:/state` to be saved, they should be renamed to `$:/temp`.

Renaming things in core templates isn't very practical. What's more important is what takes precedence: the more exotic user who wants to preserve a state or one who wouldn't expect that? I think the core should cater for the former and give the latter a simple way to preserve what he or she wishes to.

**Comment by pmario** [comment]
>If state variables should not be saved, they need to be prefixed with $:/temp. Everything else should be saved. If there are cases where we don't want $:/state to be saved, they should be renamed to $:/temp

This is kind of a "sarcastic" introduction to the next paragraph. 

>There are different opinions which elements should use $:/state and $:/temp and there will be no possibility to have the right default settings for everyone. So what ever is the default it will be wrong.

I wanted to point out, that, what ever we do, we will be wrong. There's your opinion and there is mine. With states, it seems we have an "opposite" standpoint. ... My proposal should and imo will make both of our standpoints possible in a backwards compatible way. 

I modified my view templates, to make the tiddler info area persistent. I'm removing all dropdowns and style them as reveals. Some of them use persistent states. Some of them are tabs in the tiddler info area now.

**Comment by tobibeer** [comment]
> This is kind of a "sarcastic" introduction to the next paragraph.

I'm slow on that. ^_^

> I modified my view templates, to make the tiddler info area persistent.
> I'm removing all dropdowns and style them as reveals. 

Funny you'd mention the info-panel, seeing as how it has this curious habit of resetting its state, unlike other components.

Is really what I mean with "states are temporary". To use states as a paradigm to design a view — rather than how a view behavies — imho, is not a very good / stable design choice.

**Comment by pmario** [comment]
>Is really what I mean with "states are temporary". 

Sometimes!

eg: Selected tabs use state variables, to stay selected. I don't want them to change if I save and reload the TW, while I'm creating content.  ... I may want to reset them in read only "kiosk" mode. But the "kiosk mode" tab is not necessarily the default tab, which usually is the first one.

**Comment by pmario** [comment]
@Jermolene -- Can be closed --> I did just try this one with v5.3.1 - The search box input is not saved anymore.

**Comment by Jermolene** [comment]
Thank you @pmario
